File: MetaGeta/MetaGeta.Browser/Browser/Browser.vb
Public Class Browser
    Private m_DimensionPath As List(Of String)
    Private m_CurrentLocation As New List(Of MGTag)
    Private m_CurrentFiles As FileSet
    Private m_DataStore As MGDataStore

    Friend Sub New(ByVal ds As MGDataStore, ByVal path As String)
        m_DataStore = ds
        m_DimensionPath = New List(Of String)(path.Split("/"c))
        UpdateCurrent()
    End Sub

    Public Function GetCurrentView() As TagView
        Return New TagView(m_CurrentFiles.GetAllTags(NextTagName()))
    End Function

    Public Function GetFileView() As FileView
        Return New FileView(m_CurrentFiles)
    End Function

    Public Sub Pop()
        m_CurrentLocation.RemoveAt(m_CurrentLocation.Count - 1)
        UpdateCurrent()
    End Sub
    Public Sub Push(ByVal tag As MGTag)
        If m_CurrentLocation.Count >= m_DimensionPath.Count Then Throw New Exception()
        If NextTagName() <> tag.Name Then Throw New Exception()

        m_CurrentLocation.Add(tag)
        UpdateCurrent()
    End Sub

    Public ReadOnly Property CanPush() As Boolean
        Get
            Return m_CurrentLocation.Count = m_DimensionPath.Count
        End Get
    End Property
    Public ReadOnly Property CanPop() As Boolean
        Get
            Return m_CurrentLocation.Count <> 0
        End Get
    End Property

    Public ReadOnly Property CurrentLocation() As String
        Get
            Return (From t In m_CurrentLocation Select t.Value).JoinToString(" } ")
        End Get
    End Property



    Private Sub UpdateCurrent()
        'm_CurrentFiles = m_DataStore.GetAllFiles()
        'For Each tag As MGTag In m_CurrentLocation
        '    If (m_CurrentFiles Is Nothing) Then
        '        m_CurrentFiles = m_DataStore.GetFilesWhere(tag)
        '    Else
        '        m_CurrentFiles = FileSet.Intesect(m_CurrentFiles, m_DataStore.GetFilesWhere(tag))
        '    End If
        'Next
    End Sub

    Private ReadOnly Property NextTagName() As String
        Get
            Return m_DimensionPath(m_CurrentLocation.Count)
        End Get
    End Property

End Class

File: MetaGeta/MetaGeta.Browser/Browser/Browser.cs
// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System;
using System.Collections.Generic;
using System.Linq;
using MetaGeta.DataStore;
using MetaGeta.Utilities;

#endregion

namespace MetaGeta.Browser {
    public class Browser {
        private readonly FileSet m_CurrentFiles;
        private readonly List<MGTag> m_CurrentLocation = new List<MGTag>();
        private readonly List<string> m_DimensionPath;

        private MGDataStore m_DataStore;

        internal Browser(MGDataStore ds, string path) {
            m_DataStore = ds;
            m_DimensionPath = new List<string>(path.Split('/'));
            UpdateCurrent();
        }

        public bool CanPush {
            get { return m_CurrentLocation.Count == m_DimensionPath.Count; }
        }

        public bool CanPop {
            get { return m_CurrentLocation.Count != 0; }
        }

        public string CurrentLocation {
            get { return (from t in m_CurrentLocation select t.Value).JoinToString(" } "); }
        }

        private string NextTagName {
            get { return m_DimensionPath[m_CurrentLocation.Count]; }
        }

        public TagView GetCurrentView() {
            return new TagView(m_CurrentFiles.GetAllTags(NextTagName));
        }

        public FileView GetFileView() {
            return new FileView(m_CurrentFiles);
        }

        public void Pop() {
            m_CurrentLocation.RemoveAt(m_CurrentLocation.Count - 1);
            UpdateCurrent();
        }

        public void Push(MGTag tag) {
            if (m_CurrentLocation.Count >= m_DimensionPath.Count)
                throw new Exception();
            if (NextTagName != tag.Name)
                throw new Exception();

            m_CurrentLocation.Add(tag);
            UpdateCurrent();
        }


        private void UpdateCurrent() {
            //m_CurrentFiles = m_DataStore.GetAllFiles()
            //For Each tag As MGTag In m_CurrentLocation
            //    If (m_CurrentFiles Is Nothing) Then
            //        m_CurrentFiles = m_DataStore.GetFilesWhere(tag)
            //    Else
            //        m_CurrentFiles = FileSet.Intesect(m_CurrentFiles, m_DataStore.GetFilesWhere(tag))
            //    End If
            //Next
        }
    }
}